File: src/main/java/org/cerion/stocklist/charts/VolumeChart.java
package org.cerion.stocklist.charts;

import org.cerion.stocklist.arrays.FloatArray;
import org.cerion.stocklist.arrays.ValueArray;
import org.cerion.stocklist.functions.IOverlay;
import org.cerion.stocklist.functions.ISimpleOverlay;
import org.cerion.stocklist.functions.IPriceOverlay;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class VolumeChart extends StockChart {

    public boolean logScale = false;

    @Override
    public List<IDataSet> getDataSets() {
        List<IDataSet> result = new ArrayList<IDataSet>();

        DataSet data = new DataSet(getPriceList(logScale).getVolume(), "Volume", colorBlack());
        data.setLineType(LineType.BAR);
        result.addAll(Arrays.asList(data) );

        result.addAll(getOverlayDataSets());
        return result;
    }

    private List<IDataSet> getOverlayDataSets() {
        resetNextColor();
        List<IDataSet> result = new ArrayList<IDataSet>();
        FloatArray volume = getPriceList(logScale).getVolume();

        for(IOverlay overlay : mOverlays) {
            ISimpleOverlay ol = (ISimpleOverlay)overlay;

            ValueArray arr = ol.eval(volume);
            result.addAll(getDefaultOverlayDataSets(arr, overlay));
        }

        return result;
    }
}

File: src/main/kotlin/charts/VolumeChart.kt
package org.cerion.stocks.core.charts

import org.cerion.stocks.core.PriceList
import org.cerion.stocks.core.arrays.FloatArray
import org.cerion.stocks.core.functions.ISimpleOverlay
import java.util.*

class VolumeChart : StockChart() {
    var logScale = false

    override fun getDataSets(priceList: PriceList): List<IDataSet> {
        val result = ArrayList<IDataSet>()
        val volume = if(logScale) priceList.toLogScale().volume else priceList.volume

        val data = DataSet(volume, "Volume", colorBlack())
        data.lineType = LineType.BAR
        result.addAll(listOf(data))

        val sets = getOverlayDataSets(volume)
        result.addAll(sets)
        return result
    }

    private fun getOverlayDataSets(volume: FloatArray): List<IDataSet> {
        resetNextColor()
        val result = ArrayList<IDataSet>()

        for (overlay in mOverlays) {
            val ol = overlay as ISimpleOverlay

            val arr = ol.eval(volume)
            result.addAll(getDefaultOverlayDataSets(arr, overlay))
        }

        return result

    }

}
